### ISSUE TITLE
Problem for help

### ISSUE THREAD

**Issue by xuyin218** [Yusufma03/pfrnns#5]
Hello! I'm sorry to ask for help once again. Although I have read your codes, I cannot understand some.
1、In your code, you do "obs_extractor(input_)" and "obs_extractor(input_)act_extractor(input_)", but I cannot understand why do it. And is the step necessary? If I do time-series prediction, Do I need the step too?
2、like the upper problem, in your paper, the transition update bt=f(bt-1,ut)，the observation update...，what mean about this? Is it the "State equation" and "observation equation" in Particle Filter? I do not understand the Figure 2 about the belief update, and I do not know where in the code corresponding to the transition update and observation update.
Look forward to your favourable reply, thank you!

**Comment by Yusufma03** [comment]
Hey there! 

I think I can answer these two questions altogether. In standard Bayes filters, including particle filters, we have two update steps:
- prediction step: given a state s_t, we take in a control a_t, and try to predict the next state s_{t+1};
- observation step: given a new observation o_t, we update the predicted state s_{t+1} to a posterior state s'_{t+1}

Please refer to the Background section in the appendix of my paper for more details. 

In PF-RNNs, as we don't know what is exactly the control and observation of the dynamic model, so we learn to extract them from the input. Fig. 2 simply illustrates that we are gonna split the input x_t, into the observation o_t and control a_t. In my code, this corresponds to the obs_extractor and act_extractor. I fed the output from the two extractors to the transition function and observation function afterward. 

I think this is quite useful in some specific tasks, e.g., robot localization, where we have both control and observations embedded in the input. Another example is when we apply it to reinforcement learning (see my [ICLR paper](https://github.com/Yusufma03/DPFRL)), we can learn the underlying control of the environment using this technique. However, for your case, in time series prediction, I think it might not be necessary as it is unclear what a control / observation is in that case.

I think you can have a try on both, with or without those two extractors, and choose the structure that gives the best performance. For the PF-RNNs without two extractors, you could simply embed the inputs to a latent vector using a fully connected layer, then feed this vector to both the transition function and the observation function. Here, basically you assume both the transition function and the observation function are identity functions.

Let me know if this addresses your question.

**Comment by xuyin218** [comment]
> Hey there!
> 
> I think I can answer these two questions altogether. In standard Bayes filters, including particle filters, we have two update steps:
> 
> * prediction step: given a state s_t, we take in an action a_t, and try to predict the next state s_{t+1};
> * observation step: given a new observation o_t, we update the predicted state s_{t+1} to a posterior state s'_{t+1}
> 
> Please refer to the Background section in the appendix of my paper for more details.
> 
> In PF-RNNs, as we don't know what is exactly the control and observation of the dynamic model, so we learn to extract them from the input. Fig. 2 simply illustrates that we are gonna split the input x_t, into the observation o_t and control a_t. In my code, this corresponds to the obs_extractor and act_extractor. I fed the output from the two extractors to the transition function and observation function afterward.
> 
> I think this is quite useful in some specific tasks, e.g., robot localization, where we have both control and observations embedded in the input. Another example is when we apply it to reinforcement learning (see my [ICLR paper](https://github.com/Yusufma03/DPFRL)), we can learn the underlying control of the environment using this technique. However, for your case, in time series prediction, I think it might not be necessary as it is unclear what a control / observation is in that case.
> 
> I think you can have a try on both, with or without those two extractors, and choose the structure that gives the best performance. For the PF-RNNs without two extractors, you could simply embed the inputs to a latent vector using a fully connected layer, then feed this vector to both the transition function and the observation function. Here, basically you assume both the transition function and the observation function are identity functions.
> 
> Let me know if this addresses your question.

Thank for your explicit reply!
1、In your codes, I find the obs(from obs_extractor) to feed "att = torch.cat((h1, obs), dim=1) logpdf_obs = self.fc_obs(att)", so the "fc_obs" is observation function? then the function is used to update particle weights?
2、For the time-series prediction, when I use  "att = torch.cat((h1, act), dim=1)", the result do not have too much different. 
3、When I did the comparison test, I cannot get you result 4.93 of GRU in your paper(NASDAQ), and find the advantage of PF-RNN is not obvious. I think it must be somewhere wrong in my codes. If you have the GRU codes, could you send me?(xuyin218@mail.ustc.edu.cn)

**Comment by Yusufma03** [comment]
>  1、In your codes, I find the obs(from obs_extractor) to feed "att = torch.cat((h1, obs), dim=1) logpdf_obs = self.fc_obs(att)", so the "fc_obs" is observation function? then the function is used to update particle weights?

Yes. 

>  2、For the time-series prediction, when I use "att = torch.cat((h1, act), dim=1)", the result do not have too much different.

Ya, I think for time series prediction, there is no clear definition to the control. Using this function might not be necessary. 

>  3、When I did the comparison test, I cannot get you result 4.93 of GRU in your paper(NASDAQ), and find the advantage of PF-RNN is not obvious. I think it must be somewhere wrong in my codes. If you have the GRU codes, could you send me?(xuyin218@mail.ustc.edu.cn)

I'll email you. Issue closed.
